Orm/Xtensive.Orm/Orm/Linq/Translator.Queryable.cs

```diff
@@ -1572,5 +1570,11 @@ private Expression VisitContainsNone(Expression setA, Expression setB, bool isRo
         return VisitAll(setB, lambda, isRoot);
       }
     }
+
+    private bool IsKeyCollection(Type localCollectionType)
+    {
+      return (localCollectionType.IsArray && localCollectionType.GetElementType() == typeof(Key))
+        || typeof(IEnumerable<Key>).IsAssignableFrom(localCollectionType);

```

Comment:
`typeof()` operator is not free it implies system call.
For the future it is good idea to cache known types into static readonly fields.
Please see https://github.com/DataObjects-NET/dataobjects-net/pull/51/files#diff-216f8b7482a2fe8c2cc2e71cf76eb4a7
or
https://github.com/DataObjects-NET/dataobjects-net/pull/51/files#diff-5c9e5c3c17d64a024a021916060f637d
or
https://github.com/DataObjects-NET/dataobjects-net/pull/51/files#diff-43a94c13d7718af1d89d49a6a40b0374R25

for the reference.
We can discuss this approach.

PS: At the same time `typeof` makes sense for generic type arguments or in the attribute declarations.

Reply:
Yes, I know that but now there is no place for putting such well-known types yet. I'd vote for ```WellKnownTypes``` static class (similar to ```Xtensive.Orm.Linq.WellKnownMembers```) and replace all ```typeof```s across all performance sensitive places. But I feel that it should be in a separate PR.